### ISSUE TITLE
Query

### ISSUE THREAD

**Issue by RWOverdijk** [benwinding/react-admin-firebase#20]
I decided to take a look at the code of this plugin and noticed you opted for local filtering.

The cool thing about the firebase sdk is that it already includes features to do that for you. Here's a link to the docs for simple queries: https://firebase.google.com/docs/firestore/query-data/queries#simple_queries

**Comment by benwinding** [comment]
True, that's possible, and probably a good optimisation feature to add to the options. The reason local filtering was originally chosen is that this package is naively pulling the whole collection anyway so might as well filter on the client too.

TODO:
- Pagination from the server (in options)
- Filtering from the server (in options)

**Comment by RWOverdijk** [comment]
I understand that for sure. It's nice to use Query because it also makes your onSnapshot aware of your filters. On top of that updates will probably be visible sooner. I noticed some jumpy data because local redux state and firestore state mismatched for a bit.

**Comment by benwinding** [comment]
This relates to pr: #28 right?

**Comment by RWOverdijk** [comment]
@benwinding A bit yeah. #28 is for the base query though, as a faster approach to the base path. Also, I needed this to prevent errors. App kept crashing because the user only has access to their own documents, but the current implementation doesn't discriminate.

**Comment by marknelissen** [comment]
FYI and as inspiration, this is my current decorator for the dataProvider from this library to do server side sorting, filtering and pagination.
```
const applyParams = async (query: Query, resourceName: string, type: string, { pagination, filter, sort }: QueryParams) => {
    let q = query;
    const filterKeys = filter ? Object.keys(filter) : [];
    if (filter) {
        filterKeys.forEach(fieldName => q = q.where(fieldName, "==", filter[fieldName]));
    }
    if (sort && sort.field && sort.field !== 'id' && !filterKeys.includes(sort.field)) {
        //@ts-ignore
        q = q.orderBy(sort.field, sort.order && sort.order.toLowerCase())
    }
    if (pagination) {
        if (pagination.perPage) {
            q = q.limit(pagination.perPage);
            if (pagination.page > 1) {
                const previousPage = {
                    filter,
                    pagination: {
                        ...pagination,
                        page: pagination.page - 1,
                    },
                    sort,
                };
                const paginationKey = JSON.stringify({ ...previousPage, resourceName, type });
                let ref = paginationRefs[paginationKey];
                if (!ref) {
                    await selfDataProvider(type, resourceName, previousPage);
                    ref = paginationRefs[paginationKey];
                }
                q = q.startAfter(ref);
            }
        }
    }
    return q;
}

const extractData = (result: QuerySnapshot, resourceName: string, type: string, pagination?: Pagination, filter?: Filter, sort?: Sort) => {
    const paginationKey = JSON.stringify({ filter, pagination, resourceName, sort, type });
    paginationRefs[paginationKey] = result.docs[result.docs.length - 1];
    return result.docs.map(parseFireStoreDocument);
}

const userDataProvider = (dataProvider: DataProvider) => async (type: string, resourceName: string, params: any) => {
    switch (type) {
        case GET_LIST: {
            let query: Query = db.collection(resourceName);
            query = await applyParams(query, resourceName, type, params);
            const result = await Promise.all([query.get(), metaRef.doc(resourceName).get()]);
            const data = extractData(result[0], resourceName, type, params.pagination, params.filter, params.sort);
            const meta = result[1].data();
            return {
                data,
                total: (meta && meta.total) || result[0].size,
            };
        }
        case GET_MANY_REFERENCE: {
            let query = db.collection(resourceName).where(params.target, '==', params.id);
            query = await applyParams(query, resourceName, type, params);
            const result = await query.get();
            const data = extractData(result, resourceName, type, params.pagination, params.filter, params.sort);
            return {
                data,
                total: result.size,
            };
        }
        default:
            return await dataProvider(type, resourceName, params);
    }
}

const selfDataProvider = userDataProvider(firebaseDataProvider);
```
It does come with the limitation of only exact filter matches for searches. 

It also needs a meta collection, containing a document per collection accessible through the data provider, to retrieve the total elements in a collection, which is updated with onCreate and onDelete firebase firestore function triggers.

**Comment by benwinding** [comment]
Hi @marknelissen,

Thanks for sharing your solution! 

It's quite annoying that Firestore currently has no way of determining the _size_ of a collection, which means you have to track that yourself in a meta collection, like you mentioned. And update that meta collection with firebase functions :disappointed: 

It's also hard to get fuzzy matching in Firestore, as they only allow exact matches  :disappointed: 

All this overhead and limitations are signals that Firestore is not equipped for searching. [They actually reccommend using Algolia](https://firebase.google.com/docs/firestore/solutions/search), which also has a bit of overhead in implementation...

Anyway, thanks again for Sharing it could be a good additional provider decorator that could be added to the library :+1: 

Cheers,
Ben

**Comment by marknelissen** [comment]
The fuzzy matching is a bigger issue than the size. It's only the React Admin implementation of paging that requires it. The solution is to replace the current paging List component with one with a "loading more" button, which fits better with the paging possibilities of firestore.

The closest solution, without Algolia or ElasticSearch, are the ones based on array_contains. But this requires another triggered function to update the array that contains all possible matches.

What also doesn't play nice with the current React Admin GUI, is that sorting on a field removes all documents from the result that don't have the field. I've yet to find a solution for that one.

I really like the work you have done on this library. However, if you are working on a normal sized data set, it doesn't play very well with the pricing model. You really should be more frugal with the document reads, unless you don't mind bugdet. 😉

**Comment by LaszloDev** [comment]
Firebase restricts loading a collection list (GET_LIST) if a security rule setting for one document of that list could end up being `allow read: if ...(false)`. So loading a list in such a setting requires to set the query/filter that so security rule of one document can be false.

Does this relate to the topic here?

See Firestore documentation: Rules are not filters
> Once you secure your data and begin to write queries, keep in mind that security rules are not filters. You cannot write a query for all the documents in a collection and expect Cloud Firestore to return only the documents that the current client has permission to access.
https://firebase.google.com/docs/firestore/security/rules-conditions?authuser=1#rules_are_not_filters

------
I'm currently working on an app which does have a "users" directory, and the user can decide if he whats to be listed or not.

To prevent that the data of a user, who do not what to be listed, still can not be loaded via the Firebase API, I set up the following security rule (simplified):
```
 match /users/{userID} { 
      allow read: if resource.data.directory_show 
}
```

So I would like to set a permanent filter on my list property  `<List {...props} filter={{ directory_show: true }}>`. The filter should be used to build the Firestore query in alignment which the security rules to be able to load the list.
Right now I do get an error as of missing permissions, but the query does work with the right filter set.

**Comment by marknelissen** [comment]
As I understand it, it is the same issue that made the OP investigate the query behaviour of the library indeed.

The more I advance with my own project, the more I'm convinced it will be impossible to find a generic solution that is compatible with the react-admin dataProvider API.

Maybe the focus of this library should shift from providing a fully generic dataProvider, to just provide the building blocks to more easily craft your own dataProvider, with some sensible default (configurable) behaviours.

**Comment by LaszloDev** [comment]
I do need to have a deeper look myself, but how do the other dataProvider for e.g. a REST API do separate between API query/filter calls and local search? As from what I understand this is the problem here.
Maybe you could clarify a bit more what your concerns are.

**Comment by marknelissen** [comment]
I mainly have experience with the SimpleRestDataProvider, which just delegates everything to the backend call. The only reason we even need to consider client side filtering, is because the Firestore search API is too limited to handle all filtering cases.

Some of the issues:

- You can only query for a specific page if you know the last element of the previous page
- You can only do exact or range filtering on an attribute, no 'like' or 'contains' or other partial match which is important for user searches on string attributes
- Your sort needs to be on the same attribute on which you do the range filtering
- Sorting on an attribute automatically filters on the same attribute, removing from the query result all documents that don't contain the attribute
- Any query that *could* match a document that user cannot see, is simply refused, instead of just pretending the record does not exist (eg. you cannot just fetch a full collection and do client side filtering if you have any access rule defined on the documents of the collection)

The react-admin list views are clearly set up with the server-side filtering in mind, since that is the only efficient way to combine paging, sorting and filtering, without fetching too much data client side. This is actually a good thing, but it doesn't match well with all backend systems. React-admin assumes a backend API that allows to directly fetch a specific page, while the Firestore paging logic is oriented for continuous scroll/loading list views, where you always start on the first page and sequentially load the following pages.

On top of that, the Firebase pricing model is too punishing on *unnecessary* reads, which you obviously get with client side filtering.

**Comment by LaszloDev** [comment]
So one idea would be to have a separated Firebase function which does take the search query, does a server side filtering and returning the matched documents data or their keys which does can be used to propagate the list view.

As said I still didn't had time to look at the search in specific. My main concern right now is the user roles and permissions which is related to your last point: Getting a list of a collection which does have restricted documents for the requesting user.

First step for this is my pull request https://github.com/benwinding/react-admin-firebase/pull/36 which does enable the usage of `react-admin`s permissions. The Firebase custom claims are used in my security rules to restrict access to the collections. With this we do know, by the users roles, which collection he is allowed to access.

In specific:
A user with the role `member` is only allowed to read documents in the collection `/events` which does have the field `{ state: published }`

**Problem** right now is that `react-admin-firebase` tries to get the hole `/events`-collection but the Firebase security rules in this case do not check every document on a `list`-read, but deny the complete request as one document in the collection could be not readable for the user.

To get around this we do need to specify our Firebase query to only include documents of the `/events`-collection which does have `{ state: 'published' }`.

`<List {...props} filter={{ state: 'published' }} />`

My solution here would be that every permanent filter set to the `<List />`-component would be used in the Firebase query. By this we can, with the help of the `permissions`-property, set different filter per (user) group.
The result set would be smaller helping with the reads/pricing and a first set to server side search with Firebase.

I will have a deeper look at the logic behind the List/Filter and it might be that we need a Custom List Filter component for that, but that should be doable too.

I will open another ticket about that soon and see if I can provide a pull request. But your opinion would be good to hear.

**Comment by marknelissen** [comment]
@LaszloDev , why would you implement the query logic in a function? In terms of pricing, the reads are billed in the same manner as a query from the client. You also loose the security of Firebase, since the functions operate as an admin, and their reads are not affected by the security rules you define. Beyond that, the function is not capable of doing something the client can't with regards to queries.

**Comment by RWOverdijk** [comment]
I mentioned it before, but I do like repeating myself 😄 Why not just allow base queries? That way everyone is happy. The default fetches all the things and filters what it can filter using the query, and the rest in memory. So if you need to only get documents where a specific field (or collection) matches a query you already know, you pass that to the resource.

Is it perfect? No. But it's firestore, you use it when you need it. If you don't like it, it means the tool doesn't fit your needs and you should look at alternative solutions, like prisma or building your own api.

Or something like that.

**Comment by LaszloDev** [comment]
I think we are talking about the same, regards the base queries. Did had a look at your pull request and I would favor an option where the _permanent list filter_  (`<List filter={{ "state": "published" }} />`) is used for building the query. There by the Firestore query can be set dynamically based on the _permissions_ of the logged in user.

Will submit in a merge request with my proposal tomorrow. Just finished a working version and have to polish it a bit more before.

**Comment by benwinding** [comment]
Thankyou all for your input. I like the idea of a _base_ or _permanent_ filter which is applied to the list in the background (possibly changed by permissions). This would reduce reads considerably and narrow down the local client filtering.

Another idea could be toggling **Local** (fuzzy matching) or **Server** (exact match) for the search box on the list.

It's hard to accept that there's no way to have fuzzy matching on the server-side filtering...

**Comment by marknelissen** [comment]
There is a soft variant of the fuzzy matching, in the form of a startWith formula, as described in the first answer here: https://stackoverflow.com/questions/46642641/sql-like-operator-in-cloud-firestore .
`ref.collection('user').orderBy('name').startAt(name).endAt(name+'\uf8ff')`
or 
`ref.collection('user').where('name','>=',name).where('name','<=',name+'\uf8ff')`

The main downside is that this claims the single sort field, meaning you cannot combine this with a sort on another field.

This is my main blocking point in making a full server-side query with fuzzy matching. The only solution would be to do client-side sorting, but that also implies client side paging.

As long as you only do exact matches, my solution above handles everything in the Firestore query itself.

There is another solution (besides the Algolia route of the official documentation), by populating an array attribute on your documents with all possible text matches, and using an array_contains function. According to the documentation, it seems to me this is combinable with a sort, but I've yet to test this solution, since it relies heavily on firestore trigger functions to populate the array.

**Comment by LaszloDev** [comment]
Have a look at my pull request (https://github.com/benwinding/react-admin-firebase/pull/39) regards the "base" query / permanent filter which allows to work with restricted collections.

**Comment by LaszloDev** [comment]
Resolved due to #39
